python_modules/dagster/dagster/_core/execution/context/invocation.py

```diff
@@ -779,13 +782,158 @@ def build_op_context(
     )
 
 
+############################################
+### Direct invocation AssetExecutionContext
+############################################
+
+
+class UnboundAssetExecutionContext(AssetExecutionContext):
+    """The ``context`` object available as the first argument to a asset's compute function when
+    being invoked directly. Can also be used as a context manager.
+    """
+
+    def __init__(
+        self,
+        config: Any,
+        resources_dict: Mapping[str, Any],
+        resources_config: Mapping[str, Any],
+        instance: Optional[DagsterInstance],
+        partition_key: Optional[str],
+        partition_key_range: Optional[PartitionKeyRange],
+        assets_def: Optional[AssetsDefinition],
+    ):
+        self._op_execution_context = build_op_context(
+            resources=resources_dict,
+            op_config=config,
+            resources_config=resources_config,
+            instance=instance,
+            partition_key_range=partition_key_range,
+            partition_key=partition_key,
+            mapping_key=None,
+            _assets_def=assets_def,
+        )
+
+        self._cm_scope_entered = False
+
+    def __enter__(self):
+        self._cm_scope_entered = True
+        return self
+
+    def __exit__(self, *exc):
+        self._exit_stack.close()
+
+    def __del__(self):
+        self._exit_stack.close()
```

Comment:
I don't see why you need to have an `exit_stack` in here. However I do believe the inner `OpExecutionContext``__enter__`, `__exit__`, and `__del__` methods will have to delegate to the underlying context in order for this to work.

Make sure to test resources that require setup and teardown to validate.